### PR TITLE
fix L1 candidate matching in HLTMuonL1TFilter

### PR DESCRIPTION
**Pull request by JanFSchulte**
(split from #37086)

Fix to the L1 candidate matching in HLTMuonL1TFilter. In this filter a match between L1 muons before and after different filters is performed. Currently the code expects the muons that are compared to be in the same collection. However, there are use cases in the current menu where the matching is performed with a copy of the L1 muon collection and thus always fails. This causes an inefficiency for low pT muons that are reconstructed only by L1-seeded reconstruction steps. Impact is mostly negligible, showing tiny efficiency improvements for high-pT muon from the Z:
image

Effect gets more visible for lower pT muons, for example in the turnon of the double-muon trigger:
image

However, the effect is significant for very low pT leptons. For the muon Bs -> mumu trigger (HLT_DoubleMu4_3_Bs_v14) we find a ~20% improvement of HLT efficiency on a sample of signal events provided by BPH.

### CODE REVIEW

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc [hunk 1]
```diff
@@ -110,7 +113,13 @@ bool HLTMuonL1TFilter::hltFilter(edm::Event& iEvent,
       MuonRef muon(allMuons, distance(allMuons->begin(allMuons->getFirstBX()), it));
 
       // Only select muons that were selected in the previous level
-      if (find(prevMuons.begin(), prevMuons.end(), muon) == prevMuons.end())
+      bool matchPrevL1 = false;
+      int prevSize = prevMuons.size();
+      for (int it2 = 0; it2 < prevSize; it2++) {
+        if (deltaR2(muon->eta(), muon->phi(), prevMuons[it2]->eta(), prevMuons[it2]->phi()) < maxDR_)
```
Comment (missirol) [inline HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:119]: General comment: there is some ambiguity in "deltaR vs deltaR2" (the parameter name suggests it's for DeltaR, while the actual cut is on DeltaR2). As done already here, the actual cut should be on DeltaR^2. My suggestion to mimic how this is handled [here](https://github.com/cms-sw/cmssw/blob/cf1af5f9671cb22fd9245813710e8ccdec100a67/HLTrigger/JetMET/interface/HLTPFJetsMatchedToFilteredJetsProducer.h#L48), reading the "maxDeltaR" value from the configuration, checking that it is positive, and then using the maxDeltaR2 value for the cut. Please also change `maxDR` to `maxDeltaR` to improve clarity.

Comment (JanFSchulte) [reply HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:124]: Ok, I have implemented it according to the suggestion.

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc [hunk 2]
```diff
@@ -110,7 +118,13 @@ bool HLTMuonL1TFilter::hltFilter(edm::Event& iEvent,
       MuonRef muon(allMuons, distance(allMuons->begin(allMuons->getFirstBX()), it));
 
       // Only select muons that were selected in the previous level
-      if (find(prevMuons.begin(), prevMuons.end(), muon) == prevMuons.end())
+      bool matchPrevL1 = false;
+      int prevSize = prevMuons.size();
+      for (int it2 = 0; it2 < prevSize; it2++) {
+        if (deltaR2(muon->eta(), muon->phi(), prevMuons[it2]->eta(), prevMuons[it2]->phi()) < maxDR2_)
+          matchPrevL1 = true;
```
Comment (missirol) [inline HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:125]:
```suggestion
        if (deltaR2(muon->eta(), muon->phi(), prevMuons[it2]->eta(), prevMuons[it2]->phi()) < maxDR2_){
          matchPrevL1 = true;
          break;
        }
```
Can use `break`.

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc [hunk 3]
```diff
@@ -38,7 +41,11 @@ HLTMuonL1TFilter::HLTMuonL1TFilter(const edm::ParameterSet& iConfig)
     //     }
     qualityBitMask_ |= 1 << selectQualitie;
   }
-
+  //make sure cut parameter for candidate matching is
+  if (maxDR_ <= 0.) {
+    throw cms::Exception("HLTPFJetsMatchedToFilteredJetsProducerConfiguration")
```
Comment (missirol) [inline HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:46]:
```suggestion
    throw cms::Exception("HLTMuonL1TFilterConfiguration")
```

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc [hunk 4]
```diff
@@ -38,7 +41,11 @@ HLTMuonL1TFilter::HLTMuonL1TFilter(const edm::ParameterSet& iConfig)
     //     }
     qualityBitMask_ |= 1 << selectQualitie;
   }
-
+  //make sure cut parameter for candidate matching is
```
Comment (missirol) [inline HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:44]:
```suggestion
  //make sure cut parameter for candidate matching is strictly positive
```

**Review by missirol — COMMENTED** [review]
Here's the (short) review.

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc [hunk 5]
```diff
@@ -110,7 +118,14 @@ bool HLTMuonL1TFilter::hltFilter(edm::Event& iEvent,
       MuonRef muon(allMuons, distance(allMuons->begin(allMuons->getFirstBX()), it));
 
       // Only select muons that were selected in the previous level
-      if (find(prevMuons.begin(), prevMuons.end(), muon) == prevMuons.end())
+      bool matchPrevL1 = false;
+      int prevSize = prevMuons.size();
+      for (int it2 = 0; it2 < prevSize; it2++) {
+        if (deltaR2(muon->eta(), muon->phi(), prevMuons[it2]->eta(), prevMuons[it2]->phi()) < maxDR2_)
+          matchPrevL1 = true;
+        break;
```
Comment (missirol) [inline HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:126]:
```suggestion
        if (deltaR2(muon->eta(), muon->phi(), prevMuons[it2]->eta(), prevMuons[it2]->phi()) < maxDR2_) {
          matchPrevL1 = true;
          break;
        }
```
Careful with brackets :)

Comment (JanFSchulte) [reply HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:126]: Yikes, I am really not awake enough today to code. Fix incoming...

Comment (missirol) [reply HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:126]: @JanFSchulte , while you are at it, consider squashing into 1 commit (optional).

**Comment by missirol** [comment]
type bugfix

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc (hunk 5 above)
Comment (JanFSchulte) [reply HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc]: Sure, done.

**Comment by missirol** [comment]
please test

**Comment by missirol** [comment]
+hlt

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc [hunk 6]
```diff
@@ -110,7 +118,15 @@ bool HLTMuonL1TFilter::hltFilter(edm::Event& iEvent,
       MuonRef muon(allMuons, distance(allMuons->begin(allMuons->getFirstBX()), it));
 
       // Only select muons that were selected in the previous level
-      if (find(prevMuons.begin(), prevMuons.end(), muon) == prevMuons.end())
+      bool matchPrevL1 = false;
```
Comment (perrotta) [inline HLTrigger/Muon/plugins/HLTMuonL1TFilter.cc:121]: `deltaR2` is a rather expensive operation (ok, better than `deltaR`): I would suggest to move this check after all other, cheaper, ones

File: HLTrigger/Muon/plugins/HLTMuonL1TFilter.h [hunk 7]
```diff
@@ -38,6 +38,10 @@ class HLTMuonL1TFilter : public HLTFilter {
   /// pT threshold
   double minPt_;
 
+  /// max dRs for L1 candidate matching
+  double maxDR_;
```
Comment (perrotta) [inline HLTrigger/Muon/plugins/HLTMuonL1TFilter.h:42]: If you decide to touch the code, these and all other private class members can also be made `const`

**Comment by Martin-Grunewald** [comment]
please test

**Comment by missirol** [comment]
+hlt

**Comment by perrotta** [comment]
+1

**Pull request merged**